docs/metatable-type-functions.md

```diff
@@ -0,0 +1,79 @@
+# Metatable Type Functions
+
+## Summary
+
+Implement type functions for `getmetatable` and `setmetatable`.
+
+## Motivation
+
+### `setmetatable` Type Function
+
+There is currently no way for users to apply metatable type information to a type without the usage of `typeof()`. This isn't ideal, as it adds verbosity and boilerplate to common patterns such as object-oriented programming. For example, the following:
+
+```lua
+local clock = {}
+type Identity = typeof(setmetatable({} :: { time: number }, { __index = clock }))
+```
+
+could be reduced to:
+
+```lua
+local clock = {}
+type Identity = setmetatable<{ time: number }, { __index: clock }>
+```
+
+### `getmetatable` Type Function
+
+[Issue #1435](https://github.com/luau-lang/luau/issues/1435) is caused by a lack of special behavior for `getmetatable` when the type's metatable has a `__metatable` field. This is fixable in a variety of ways, however a type function has multiple benefits and exposes more type-level expression for users. It fits in nicely with the new solver, as type functions are a cleaner alternative to magic functions.
+
+## Design
+
+### `setmetatable` Type Function
+
+In the following code example, `Identity` should evaluate to `{ sound: string, @metatable: { __index: animal } }`:
+
+```lua
+local animal = {}
+type Identity = setmetatable<{
+	sound: string
+}, {
+	__index: animal
+}>
+```
+
+### `getmetatable` Type Function
```

Comment:
What is the semantics of `getmetatable` when the argument does not have a metatable? What about when it cannot have one, e.g. because it is neither a table nor class? What does it do when given a union of tables? Should it check if the tables all have the same metatable and return the one or error otherwise? Should it return a union of the metatables across the union?

Reply:
My general impression is that it makes the most sense to air conservatively here, we can always specify more behavior later. Possible answer: `getmetatable` will work on a table, and a class, and give you their respective metatables, but nothing else, and will respect `__metatable`.